{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "246bf566",
   "metadata": {},
   "source": [
    "# E. faecium nanopore workflow\n",
    "This notebook details the full computational workflow for the paper: \"Unmasking a Superbug: Nanopore Sequencing Reveals Chromosomal *vanB* and Resistance Plasmids in VRE *Enterococcus faecium* ST177\".\n",
    "\n",
    "It is a \"hybrid\" workflow, meaning it contains:\n",
    "1.  **Local steps** you can run directly in this notebook.\n",
    "2.  **HPC steps** that require transferring files to a cluster."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44f5bed0",
   "metadata": {},
   "source": [
    "## Before You Begin\n",
    "This notebook **will not work** unless you complete these three setup steps first."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9508f324",
   "metadata": {},
   "source": [
    "### Get the Data\n",
    "input your Nanopore data in the <code>data/raw/</code> folder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e07186d3",
   "metadata": {},
   "source": [
    "### Create the Environment\n",
    "This notebook is best run using the <code>efaecium_env</code> conda environment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f837b99c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------[ SET UP CONDA ENVIRONMENT ]-------------\n",
    "\n",
    "!conda env create -f \"../environment.yml\"\n",
    "# This creates the efaecium_env conda environment\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b51e08d",
   "metadata": {},
   "source": [
    "Activate this environment for this Jupyter kernel before running the cells below"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a66e5aff",
   "metadata": {},
   "source": [
    "### Configure Your Paths\n",
    "Edit the <code>config.yaml</code> file in the project's root directory. All local file paths in this notebook are loaded from that single config file.\n",
    "\n",
    "The HPC paths and settings are loaded from the <code>HPC_config.py</code> file located in the <code>HPC/</code> folder\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b632d3cc",
   "metadata": {},
   "source": [
    "## Imports and Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "308ece0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------[ IMPORTS ]-------------\n",
    "import yaml\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ac35941",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------[ CONFIG ]-------------\n",
    "# Get the path to the config file\n",
    "config_path = os.path.join('..', 'config.yaml')\n",
    "\n",
    "# Load configurations from the yaml file\n",
    "with open(config_path, 'r') as f:\n",
    "    config = yaml.safe_load(f)\n",
    "print(\"Configuration loaded successfully.\")\n",
    "\n",
    "# Get the raw file path\n",
    "raw_fastq_path = os.path.join('..',config['raw_fastq_path'])\n",
    "print(f\"the raw fastq path is: {raw_fastq_path}\")\n",
    "\n",
    "# Get the Fastplong output path and proccesed file path\n",
    "fastplong_output_path = os.path.join('..', config['fastplong_output_path'])\n",
    "fastplong_html_path = os.path.join(fastplong_output_path, 'fastplong_report.html')\n",
    "fastplong_json_path = os.path.join(fastplong_output_path, 'fastplong_report.json')\n",
    "print(f\"The Fastplong output path is: {fastplong_output_path}\")\n",
    "fastplong_filtered_path = os.path.join('..', config['fastplong_filtered_path'])\n",
    "print(f\"The Fastplong filtered path is: {fastplong_filtered_path}\")\n",
    "\n",
    "# Get the MultiQC input and output path\n",
    "multiqc_input_path = os.path.join('..', config['multiqc_input_path'])\n",
    "print(f\"The MultiQC input path is: {multiqc_input_path}\")\n",
    "multiqc_output_path = os.path.join('..', config['multiqc_output_path'])\n",
    "print(f\"The MultiQC output path is: {multiqc_output_path}\")\n",
    "\n",
    "# Get the Bakta database path and input genome path\n",
    "bakta_db_folder = os.path.join('..', config['bakta_db_path'])\n",
    "bakta_db_path = os.path.join(bakta_db_folder, 'db')\n",
    "print(f\"The Bakta database path is: {bakta_db_path}\")\n",
    "bakta_input_path = os.path.join('..', config['flye_assembly_path'])\n",
    "print(f\"The bakta_input_path path is: {bakta_input_path}\")\n",
    "bakta_output_path = os.path.join('..', config['bakta_output_path'])\n",
    "print(f\"The Bakta output path is: {bakta_output_path}\")\n",
    "\n",
    "# Get the AMRFinderPlus output path\n",
    "amrfinder_output_path = os.path.join('..', config['amrfinder_output_file'])\n",
    "print(f\"The AMRFinderPlus output path is: {amrfinder_output_path}\")\n",
    "\n",
    "# Get the MLST output path\n",
    "mlst_output_file = os.path.join('..', config['mlst_output_file'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75fef704",
   "metadata": {},
   "source": [
    "## Initial processing and quality control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c72d2454",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------[ FASTPLONG ANALYSIS ]-------------\n",
    "# make output directory if it doesn't exist\n",
    "!mkdir -p {fastplong_output_path}\n",
    "\n",
    "# Run Fastplong to filter and generate reports\n",
    "!fastplong \\\n",
    "  -i {raw_fastq_path} \\\n",
    "  -o {fastplong_filtered_path} \\\n",
    "  -h {fastplong_html_path} \\\n",
    "  -j {fastplong_json_path}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45a9184f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------[ INITIAL MULTIQC ]-------------\n",
    "\n",
    "!multiqc {multiqc_input_path} \\\n",
    "    --title \"Initial QC\" \\\n",
    "    --filename \"initial_QC\" \\\n",
    "    --outdir {multiqc_output_path} \\\n",
    "    --dirs --dirs-depth 2 --force"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d16553cf",
   "metadata": {},
   "source": [
    "Let's move over to the cluster to run Kraken2, Flye, and QUAST"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "929544ef",
   "metadata": {},
   "source": [
    "## High Performance Cluster: Contmination Check and Assembly \n",
    "Certain resource intensive tasks or those that take huge databases are better run on a HPC\n",
    "\n",
    "Here we will run Kraken2, Flye, and QUAST on a cluster using the supplied singularity containers. \n",
    "\n",
    "You will need to have downloaded a Kraken2 database to your HPC for that step to work, beware that this file is very large.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "905c7c54",
   "metadata": {},
   "source": [
    "1. **Configuration**: Edit the <code>HPC_config.py</code> script to match your cluster's file paths.\n",
    "2. **Transfer files**: Use scp to transfer the whole <code>scripts/HPC</code> folder as well as the input fasta file from <code>data/processed</code>\n",
    "3. **Run jobs**: Log into the HPC and execute the following in the terminal from the folder you uploaded:\n",
    "    1.  Run contamination check and assembly<code>python 00_HPC_orchestrator --kraken2 --flye</code>\n",
    "    2.  Once flye is done, run <code>python 00_HPC_orchestrator --quast</code>\n",
    "5. **Download Results**: Return to your local environment and use scp to download the entire <code>results/</code> directory from the cluster."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8149c9fa",
   "metadata": {},
   "source": [
    "## Annotation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a98dd96",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------[ MULTIQC REPORT POST HPC  ]-------------\n",
    "!multiqc {multiqc_input_path} \\\n",
    "    --title \"Post_HPC_QC\" \\\n",
    "    --filename \"Post_HPC_QC\" \\\n",
    "    --outdir {multiqc_output_path} \\\n",
    "    --dirs --dirs-depth 2 --force"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3643b7a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Before running bakta, download the database\n",
    "# The light database is ~4GB, the full database is ~83GB\n",
    "# If you have the time ans space change --type light to --type full\n",
    "!bakta_db download --output {bakta_db_folder} --type light"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4eaf01e",
   "metadata": {},
   "source": [
    "Bacta didn't like being run from a jupyter script, just run this on the terminal:\n",
    "<code>bakta --db data/bakta_database/db.full/ --output results/bakta_output/ data/processed/assembly.fasta --force</code>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4727e5df",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------[ MULTIQC REPORT WITH BAKTA ]-------------\n",
    "!multiqc {multiqc_input_path} \\\n",
    "    --title \"Bakta_QC\" \\\n",
    "    --filename \"Bakta_QC\" \\\n",
    "    --outdir {multiqc_output_path} \\\n",
    "    --dirs --dirs-depth 2 --force"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9def125",
   "metadata": {},
   "source": [
    "After running bakta, run AMRFinderPlus to reveal AMR genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "050d639f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# First download the AMRFinderPlus database into your evnironment\n",
    "!amrfinder --update "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa07e84a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------[ RUN AMRFinderPlus SEPARATELY ]-------------#\n",
    "# Run amrfinder\n",
    "!amrfinder  -n {bakta_input_path}\\\n",
    "            -O Enterococcus_faecium\\\n",
    "            -o {amrfinder_output_path}\\"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "085a8c97",
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------[ RUN MLST TO GET SEQUENCE TYPE ]-------------\n",
    "!mlst {bakta_input_path} > {mlst_output_file}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e0ca090",
   "metadata": {},
   "source": [
    "## Analysis of results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "636f8285",
   "metadata": {},
   "source": [
    "**Identifying Plasmids using PlasmidFinder 2.1**\n",
    "\n",
    "Upload assembly file from flye to https://cge.food.dtu.dk/services/PlasmidFinder/ and download the results file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "060034ee",
   "metadata": {},
   "source": [
    "**Genomic Feature Analysis Using Gemeni 2.5**\n",
    "\n",
    "The .tsv output of bakta was run through Gemini, a Large Language Model (LLM) available at https://gemini.google.com/app \n",
    "\n",
    "The model was used as an analytical tool to rapidly synthesize the large dataset and prompted to identify complex gene clusters. All identified features, operon structures, and gene groupings suggested by the LLM were subsequently manually verified against the primary Bakta annotations and cross-referenced with relevant literature.\n",
    "\n",
    "This analysis revealed many interesting genomic features, indluding a cluster of AMR genes that AMRFinderPlus missed on one of the plasmids I thought unimportant, highlighting the need for many-pronged analysis when characterizing complex genomes."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "efaecium_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
